Fix #11325 Content-Length checks

gregw opened the pull request:
Fix #11325 by:
 + removing content-length tracking from ServletContextResponse
 + adding a setContentLengthSet field to HttpOutput

gregw committed:
Fix #11325 Content-Length checks

Fix #11325 by:
 + removing content-length tracking from ServletContextResponse
 + adding a setContentLengthSet field to HttpOutput

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java:
```diff
@@ -174,6 +175,34 @@ public long getWritten()
+     * similar handlers are used.
+     * @param len The content-length as set by the application.
+     */
+    public void setContentLengthSet(long len)
```
Rename to `setApplicationContentLength(long)`.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java:
```diff
@@ -174,6 +175,34 @@ public long getWritten()
+     */
+    public boolean isContentIncomplete()
+    {
+        return _contentLengthSet >= 0 && _written < _contentLengthSet;
```
I'd read the field into a local variable, and use the local twice, rather than the field twice.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java:
```diff
@@ -352,7 +381,7 @@ public void complete(Callback callback)
                 case PENDING: // an async write is pending and may complete at any time
                     // If this is not the last write, then we must abort
-                    if (_servletChannel.getServletContextResponse().isContentIncomplete(_written))
+                    if (_contentLengthSet >= 0 && _written < _contentLengthSet)
```
Duplicate logic? Can't you call `isContentIncomplete()`?

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java:
```diff
@@ -733,7 +762,15 @@ public void write(byte[] b, int off, int len) throws IOException
-            last = _servletChannel.getServletContextResponse().isAllContentWritten(written);
+
+            // Is this the last write due to content-length?
+            if (_contentLengthSet >= 0)
+            {
+                if (_written > _contentLengthSet)
+                    throw new IllegalStateException("too much content written");
+                last = written == _contentLengthSet;
+            }
```
This block is repeated 3 times, can be factored out into a method?

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java:
```diff
@@ -1096,6 +1149,7 @@ else if (crlf != null && crlf.hasRemaining())
      * @param content The whole content to send
      * @throws IOException if the send fails
      */
+    @Deprecated(forRemoval = true, since = "12.1.0")
```
Must add also the `@deprecated` javadoc tag.
Why are they deprecated?

gregw replied:
Because they are no longer used.   ResourceService is now written against core API

sbordet replied:
But applications were able to cast to `HttpOutput` and use these methods, for example if they had a `ByteBuffer` and did not want to use the `OutputStream` APIs.

Do we offer an alternative? If so, should be documented in the javadocs.

gregw replied:
The alternative is to use the ResourceService (now in javadoc).    Also I doubt any applications were down casting  like this.

joakime replied:
> Also I doubt any applications were down casting like this.

Our own FastFileServer demo/embedded does that.
And it has been copied to other projects for years now.

Some examples of applications down casting like that ...

* https://github.com/hamann/pedestal/blob/ecae8023662b79b95e7b4df6df9de7c130ae0aff/jetty/src/io/pedestal/http/jetty/container.clj#L19-L31
* https://github.com/apache/ignite-extensions/blob/a51fb6d0e9c1129bcf190990687134bea1efb90a/modules/mesos-ext/src/main/java/org/apache/ignite/mesos/resource/ResourceHandler.java#L125-L127
* https://github.com/haoxianrui/spring-framework/blob/e5163351c47feb69483e79fa782eec3e4d8613e8/spring-web/src/main/java/org/springframework/http/server/reactive/JettyHttpHandlerAdapter.java#L128-L131
* https://github.com/cognitedata/bazel-maven-proxy/blob/4360308592b4a4dd4165482281ecbfd08a257636/server/src/main/java/com/salesforce/bazel/maven/proxy/server/MavenProxyServlet.java#L375-L392

gregw replied:
Hmmmm then we'd better add tests for them.

gregw replied:
@joakime I've updated HttpOutputTest for EE10 and EE11

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiResponse.java:
```diff
@@ -350,12 +350,18 @@ public void setContentLengthLong(long len)
+        {
+            getServletChannel().getHttpOutput().setContentLengthSet(len);
         if (len > 0)
             getResponse().getHeaders().put(HttpHeader.CONTENT_LENGTH, len);
-        else if (len == 0)
+        else
             getResponse().getHeaders().put(HttpFields.CONTENT_LENGTH_0);
```
Wrong indentation?

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextResponse.java:
```diff
@@ -586,21 +505,17 @@ public boolean onRemoveField(HttpField field)
             if (field.getHeader() == null)
                 return true;
-            switch (field.getHeader())
+            if (Objects.requireNonNull(field.getHeader()) == HttpHeader.CONTENT_TYPE)
```
`Objects.requireNonNull()` is not necessary, as the null check is 2 lines above.

No more case for `CONTENT_LENGTH`?

gregw replied:
The content length recorded here is at the context level, which could be before or after compression, depending on configuration.  We can do all the required checks at HttpOutput level for application writes/completion and HttpChannel level for network writes/completion

sbordet reviewed (changes requested):
Comments made for ee10, but same comments apply to ee11.

gregw committed:
Fix #11325 Content-Length checks

Fix #11325 by:
 + removing content-length tracking from ServletContextResponse
 + adding a setContentLengthSet field to HttpOutput

gregw committed:
Fix #11325 Content-Length checks

Fixed flaky tests due to thread local context left set.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -1053,6 +1056,7 @@ public boolean handle(Request request, Response response, Callback callback) thr
             return true;
 
         // Past this point we are calling the downstream handler in scope.
+        Context lastContext = getCurrentContext();
```
@sbordet this is an unrelated change that was causing flaky tests as it was leaving the ThreadLocal context set on the main test thread!

gregw committed:
Fix #11325 Content-Length checks

Fix #11325 by:
 + removing content-length tracking from ServletContextResponse
 + adding a setContentLengthSet field to HttpOutput
 + added HttpOutputTest for EE10/EE11
